services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -894,12 +894,12 @@ protected Response exchangeClientToClient(UserModel targetUser, UserSessionModel
 
         event.session(targetUserSession);
 
-        AuthenticationManager.setClientScopesInSession(authSession);
-        ClientSessionContext clientSessionCtx = TokenManager.attachAuthenticationSession(this.session, targetUserSession, authSession);
+        AuthenticationManager.setClientScopesInSession(authSession, targetClient);
```

Comment:
This feels so wrong. So token exchange uses client scopes of the target client, not the initiating client, to create the token. However, that strictly means you are creating a token on behalf of the target client, not a token that is aimed at the target client?

Reply:
The token should be build to be used by the targeted audience, not the requesting client. So that any scope included should be processed in accordance with the audience constraints/settings.